hwcomposer/hwcomposer.cpp

```diff
@@ -533,26 +533,33 @@ static int hwc_set(struct hwc_composer_device_1* dev,size_t numDisplays,
             continue;
         }
 
+        const hwc_rect_t frame = fb_layer->displayFrame;
+        const int width = frame.right - frame.left;
+        const int height = frame.bottom - frame.top;
+
         struct window *window = NULL;
         std::string layer_name = pdev->display->layer_names[layer];
 
         if (active_apps == "Waydroid") {
             // Show everything in a single window
             if (pdev->windows.find(active_apps) == pdev->windows.end()) {
-                pdev->windows[active_apps] = create_window(pdev->display, pdev->use_subsurface, active_apps, "0");
+                pdev->windows[active_apps] = create_window(pdev->display, pdev->use_subsurface, active_apps, "0", width, height);
                 property_set("waydroid.open_windows", std::to_string(pdev->windows.size()).c_str());
             }
             window = pdev->windows[active_apps];
         } else if (!pdev->use_subsurface) {
             if (single_layer_tid.length()) {
                 if (pdev->windows.find(single_layer_tid) == pdev->windows.end()) {
-                    pdev->windows[single_layer_tid] = create_window(pdev->display, pdev->use_subsurface, single_layer_aid, single_layer_tid);
+                    pdev->windows[single_layer_tid] = create_window(pdev->display, pdev->use_subsurface, single_layer_aid, single_layer_tid, width, height);
                     property_set("waydroid.open_windows", std::to_string(pdev->windows.size()).c_str());
                 }
                 window = pdev->windows[single_layer_tid];
+
                 // Window is closed, don't bother
-                if (!window->isActive)
+                if (window && !window->isActive) {
+                    free(window);
```

Comment:
You're not removing the window from the list, which leads to crash. Also, I'd rather use `destroy_window` instead of freeing directly.